### PR TITLE
take-until and other coll fns

### PR DESCRIPTION
**Pull request by joel7t**
Here are some functions that I thought might belong in utilis.

The important one is `take-until`.  It takes from a collection up to and including the first item for which the predicate returns true.

`partition-all-reverse` is not an ideal implementation (not lazy, no transducer), so if preferred, I'm content to leave it out for now.  Instead of the trailing partitions being of smaller size, this version puts the smaller partitions at the front.  Another partition-all style that I've used in the past is `partition-all-symmetric` which has both leading and trailing partitions of smaller sizes.

The others are added for discussion: `llast`, `second-last`, and `drop-take`.

### CODE REVIEW

**Review by ct7t — COMMENTED** [review]
Carrying on from this morning's conversation; `second-last`, `llast`, and `take-until` all jump out to me as nice complements of `second`, `ffirst`, and `take-while`.

`drop-take` and `partition-all-reverse` are less obvious to me as universal fns.

**Comment by joel7t** [comment]
Cool.  When you split them up like that @ct7t , I see that I use `drop-take` and `partition-all-reverse` in time series situations, so I may eventually find a home for them in another library.

**Comment by joel7t** [comment]
Which implementation of `second-last` is better?

Current:
```
(defn second-last
  "Returns the second last item in the collection."
  [coll]
  (last (butlast coll)))
```

Alternate:
```
(defn second-last
  "Returns the second last item in the collection."
  [coll]
  (if (next (next coll))
    (recur (next coll))
    (first coll)))
```

I'm pretty sure the alternate is the better implementation.  Thoughts?

**Comment by tom7t** [comment]
> Which implementation of `second-last` is better?
> 
> Current:
> 
> ```
> (defn second-last
>   "Returns the second last item in the collection."
>   [coll]
>   (last (butlast coll)))
> ```
> 
> Alternate:
> 
> ```
> (defn second-last
>   "Returns the second last item in the collection."
>   [coll]
>   (if (next (next coll))
>     (recur (next coll))
>     (first coll)))
> ```
> 
> I'm pretty sure the alternate is the better implementation. Thoughts?

The first one reads a bit easier, but presumably the alternate's implementation might have some performance or memory gains? If there's performance gains for the second it would be worth doing.

**Comment by sandhu** [comment]
The approach is going to depend on the type of collection. If it's a vector you're better off just looking up the item by index. For anything else, the second option is likely faster and will take less memory.

**Comment by sandhu** [comment]
Since we're in benchmarking mode, might also be worth testing `(second (reverse coll))`

**Comment by sandhu** [comment]
Also `(first (take-last 2))`, which will be more memory efficient than `butlast`, but might not be any faster.

**Comment by joel7t** [comment]
Is this sufficient for benchmarking? `(bench (second-last-1-butlast (range 10000000)))`  It's realizing the collection every time.  Should I try a different collection size or a different collection type?

### Results
second-last-1-butlast: 1.551179 sec
second-last-2-next: 658.449352 ms
second-last-3-nth: 338.720296 ms
**second-last-4-reverse: 225.960119 ms**
second-last-5-take-last: 442.353522 ms

### Implementations
```clojure
(defn second-last-1-butlast
  [coll]
  (last (butlast coll)))

(defn second-last-2-next
  "Returns the second last item in the collection."
  [coll]
  (if (next (next coll))
    (recur (next coll))
    (first coll)))

(defn second-last-3-nth
  [coll]
  (let [n (count coll)]
    (nth coll (- n 2))))

(defn second-last-4-reverse
  [coll]
  (second (reverse coll)))

(defn second-last-5-take-last
  [coll]
  (first (take-last 2 coll)))
```
### Benchmark Outputs
```clojure
involo.flight-planner> (time (bench (second-last-1-butlast (range 10000000))))
Evaluation count : 60 in 60 samples of 1 calls.
             Execution time mean : 1.551179 sec
    Execution time std-deviation : 170.007480 ms
   Execution time lower quantile : 1.466604 sec ( 2.5%)
   Execution time upper quantile : 1.901676 sec (97.5%)
                   Overhead used : 12.434079 ns

Found 7 outliers in 60 samples (11.6667 %)
	low-severe	 3 (5.0000 %)
	low-mild	 4 (6.6667 %)
 Variance from outliers : 73.7805 % Variance is severely inflated by outliers
"Elapsed time: 108954.444602 msecs"
nil
involo.flight-planner> (time (bench (second-last-2-next (range 10000000))))
Evaluation count : 120 in 60 samples of 2 calls.
             Execution time mean : 658.449352 ms
    Execution time std-deviation : 34.728713 ms
   Execution time lower quantile : 624.779561 ms ( 2.5%)
   Execution time upper quantile : 740.249972 ms (97.5%)
                   Overhead used : 12.434079 ns

Found 3 outliers in 60 samples (5.0000 %)
	low-severe	 3 (5.0000 %)
 Variance from outliers : 38.5022 % Variance is moderately inflated by outliers
"Elapsed time: 93364.844715 msecs"
nil
involo.flight-planner> (time (bench (second-last-3-nth (range 10000000))))
Evaluation count : 180 in 60 samples of 3 calls.
             Execution time mean : 338.720296 ms
    Execution time std-deviation : 12.398976 ms
   Execution time lower quantile : 326.925598 ms ( 2.5%)
   Execution time upper quantile : 374.548813 ms (97.5%)
                   Overhead used : 12.434079 ns
                                                          
Found 7 outliers in 60 samples (11.6667 %)
	low-severe	 2 (3.3333 %)
	low-mild	 5 (8.3333 %)
 Variance from outliers : 23.7853 % Variance is moderately inflated by outliers
"Elapsed time: 75572.355594 msecs"
nil
involo.flight-planner> (time (bench (second-last-4-reverse (range 10000000))))
Evaluation count : 240 in 60 samples of 4 calls.
             Execution time mean : 225.960119 ms
    Execution time std-deviation : 65.617282 ms
   Execution time lower quantile : 183.009977 ms ( 2.5%)
   Execution time upper quantile : 394.872279 ms (97.5%)
                   Overhead used : 12.434079 ns

Found 9 outliers in 60 samples (15.0000 %)
	low-severe	 4 (6.6667 %)
	low-mild	 5 (8.3333 %)
 Variance from outliers : 96.3781 % Variance is severely inflated by outliers
"Elapsed time: 68904.777021 msecs"
nil
involo.flight-planner> (time (bench (second-last-5-take-last (range 10000000))))
Evaluation count : 180 in 60 samples of 3 calls.
             Execution time mean : 442.353522 ms
    Execution time std-deviation : 11.077302 ms
   Execution time lower quantile : 430.910597 ms ( 2.5%)
   Execution time upper quantile : 473.351250 ms (97.5%)
                   Overhead used : 12.434079 ns

Found 7 outliers in 60 samples (11.6667 %)
	low-severe	 4 (6.6667 %)
	low-mild	 3 (5.0000 %)
 Variance from outliers : 12.5951 % Variance is moderately inflated by outliers
"Elapsed time: 94815.307187 msecs"
```

**Comment by joel7t** [comment]
I am suspicious of `(reverse (range 1000))`.  Why would this implementation of `last` be faster than clojure.core?

```clojure
(defn last-2-reverse
  [coll]
  (first (reverse coll)))
```

```clojure
involo.flight-planner> (time (bench (last (range 10000000))))
Evaluation count : 180 in 60 samples of 3 calls.
             Execution time mean : 442.498734 ms
    Execution time std-deviation : 8.339583 ms
   Execution time lower quantile : 431.192582 ms ( 2.5%)
   Execution time upper quantile : 463.476358 ms (97.5%)
                   Overhead used : 12.434079 ns
                                                          
Found 4 outliers in 60 samples (6.6667 %)
	low-severe	 3 (5.0000 %)
	low-mild	 1 (1.6667 %)
 Variance from outliers : 7.8242 % Variance is slightly inflated by outliers
"Elapsed time: 94967.196263 msecs"
nil
involo.flight-planner> (time (bench (last-2-reverse (range 10000000))))
Evaluation count : 240 in 60 samples of 4 calls.
             Execution time mean : 245.468453 ms
    Execution time std-deviation : 89.426248 ms
   Execution time lower quantile : 181.743593 ms ( 2.5%)
   Execution time upper quantile : 452.136372 ms (97.5%)
                   Overhead used : 12.434079 ns
"Elapsed time: 73820.079637 msecs"
```

**Comment by joel7t** [comment]
Maybe it's the transients that give `reverse` a boost over `last`... Using `(into (list) coll)`, which reverses the collection, is comparable to `reverse`.

```clojure
(defn last-3-into
  [coll]
  (first (into (list) coll)))
```

```clojure
involo.flight-planner> (time (bench (last-3-into (range 10000000))))
Evaluation count : 360 in 60 samples of 6 calls.
             Execution time mean : 251.121655 ms
    Execution time std-deviation : 59.595179 ms
   Execution time lower quantile : 196.882164 ms ( 2.5%)
   Execution time upper quantile : 376.792461 ms (97.5%)
                   Overhead used : 12.434079 ns
"Elapsed time: 106563.999765 msecs"
```

**Comment by joel7t** [comment]
Maybe it's the transients that give `reverse` a boost over `last`... Using `(into (list) coll)`, which reverses the collection, is comparable to `reverse`.

```clojure
(defn last-3-into
  [coll]
  (first (into (list) coll)))
```

```clojure
> (time (bench (last-3-into (range 10000000))))
Evaluation count : 360 in 60 samples of 6 calls.
             Execution time mean : 251.121655 ms
    Execution time std-deviation : 59.595179 ms
   Execution time lower quantile : 196.882164 ms ( 2.5%)
   Execution time upper quantile : 376.792461 ms (97.5%)
                   Overhead used : 12.434079 ns
"Elapsed time: 106563.999765 msecs"
```

**Comment by joel7t** [comment]
Here are the results with a realized input.  This lowers the benchmark by about 200ms in most cases, except for second-reverse, which had its benchmark increase!  I re-ran the second-reverse benchmark 2 more times and got 280ms and 300ms, right around 290ms.

### Results
second-last-1-butlast: 1.349807 sec (~200ms lower)
second-last-2-next: 452.628143 ms (~200ms lower)
**second-last-3-nth: 137.963497 ms** (~200ms lower)
second-last-4-reverse: 293.380874 ms (~70ms **higher**)
second-last-5-take-last: 260.454497 ms (~180ms lower)

### Benchmark Outputs
```clojure
> (def data (doall (range 10000000)))
data
> (time (bench (second-last-1-butlast data)))
Evaluation count : 60 in 60 samples of 1 calls.
             Execution time mean : 1.349807 sec
    Execution time std-deviation : 139.225976 ms
   Execution time lower quantile : 1.248574 sec ( 2.5%)
   Execution time upper quantile : 1.582838 sec (97.5%)
                   Overhead used : 12.434079 ns
                                                          
Found 7 outliers in 60 samples (11.6667 %)
	low-severe	 4 (6.6667 %)
	low-mild	 3 (5.0000 %)
 Variance from outliers : 70.4095 % Variance is severely inflated by outliers
"Elapsed time: 102698.970396 msecs"
nil
> (time (bench (second-last-2-next data)))
Evaluation count : 120 in 60 samples of 2 calls.
             Execution time mean : 452.628143 ms
    Execution time std-deviation : 20.758140 ms
   Execution time lower quantile : 434.125466 ms ( 2.5%)
   Execution time upper quantile : 500.800406 ms (97.5%)
                   Overhead used : 12.434079 ns

Found 10 outliers in 60 samples (16.6667 %)
	low-severe	 9 (15.0000 %)
	low-mild	 1 (1.6667 %)
 Variance from outliers : 31.9576 % Variance is moderately inflated by outliers
"Elapsed time: 73653.498634 msecs"
nil
> (time (bench (second-last-3-nth data)))
Evaluation count : 480 in 60 samples of 8 calls.
             Execution time mean : 137.963497 ms
    Execution time std-deviation : 1.558515 ms
   Execution time lower quantile : 136.700902 ms ( 2.5%)
   Execution time upper quantile : 141.333049 ms (97.5%)
                   Overhead used : 12.434079 ns

Found 6 outliers in 60 samples (10.0000 %)
	low-severe	 4 (6.6667 %)
	low-mild	 2 (3.3333 %)
 Variance from outliers : 1.6389 % Variance is slightly inflated by outliers
"Elapsed time: 86820.559397 msecs"
nil
> (time (bench (second-last-4-reverse data)))
Evaluation count : 300 in 60 samples of 5 calls.
             Execution time mean : 293.380874 ms
    Execution time std-deviation : 62.728895 ms
   Execution time lower quantile : 242.417644 ms ( 2.5%)
   Execution time upper quantile : 439.430683 ms (97.5%)
                   Overhead used : 12.434079 ns
"Elapsed time: 107578.604688 msecs"
nil
> (time (bench (second-last-5-take-last data)))
Evaluation count : 240 in 60 samples of 4 calls.
             Execution time mean : 260.454497 ms
    Execution time std-deviation : 5.751837 ms
   Execution time lower quantile : 255.703043 ms ( 2.5%)
   Execution time upper quantile : 274.812623 ms (97.5%)
                   Overhead used : 12.434079 ns
                                                          
Found 6 outliers in 60 samples (10.0000 %)
	low-severe	 2 (3.3333 %)
	low-mild	 4 (6.6667 %)
 Variance from outliers : 9.4660 % Variance is slightly inflated by outliers
"Elapsed time: 81933.489218 msecs"
```

**Comment by joel7t** [comment]
If we adapt this implementation to last (instead of second-last or nth-last), and compare it to clojure.core/last:

```clojure
(defn last-2-nth
  [coll]
  (let [v (if (vector? coll) coll (vec coll))
        n (count v)]
    (nth v (- n 1))))
```
On `data-rnd-nested`:
```
last: 890ms
last-2-nth: 670ms
```
On `data-rnd-vec-nested` (notice ms vs ns):
```
last: 420ms
last-2-nth: 107ns
```

File: src/utilis/coll.cljc [hunk 1]
```diff
@@ -46,22 +46,24 @@
...
+  ([coll ind]
+   (let [v (if (vector? coll) coll (vec coll))
+         n (count v)]
+     (nth v (- n ind 1))))
+  ([coll ind not-found]
+   (let [v (if (vector? coll) coll (vec coll))
+         n (count v)]
+     (if (< 0 ind n)
+       (nth v (- n ind 1))
+       not-found))))
```
Comment (sandhu) [inline src/utilis/coll.cljc:61]: Would suggest something like the following to handle code duplications and out of bounds "at both ends"
```clojure
(defn nth-last
  "Returns the nth last item in the collection.  Throws an exception if
  the index is out of bounds unless `default` is supplied."
  ([coll index]
   (nth-last coll index ::not-found))
  ([coll ind default]
   (let [v (if (vector? coll) coll (vec coll))
         n (count v)]
     (try
       (nth v (- n ind 1))
       (catch Exception e
         (if (not= default ::not-found)
           default
           (throw e)))))))```

File: src/utilis/coll.cljc [hunk 2]
```diff
@@ -46,22 +46,24 @@
...
```
Comment (sandhu) [inline src/utilis/coll.cljc:62]: Would suggest also including a `second-last` since a `second` exists in the core library

**Review by sandhu — CHANGES_REQUESTED** [review]

**Pull request merged**